### ISSUE TITLE
null_resource with local-exec provisioner does not allow sensitive variables

### ISSUE THREAD

**Issue by oliverisaac** [hashicorp/terraform#27763]
### Terraform Version

```
Terraform v0.14.6
+ provider registry.terraform.io/hashicorp/external v2.0.0
+ provider registry.terraform.io/hashicorp/null v3.0.0```
```
### Terraform Configuration Files

```terraform
variable "secret" {
  default   = "password123"
  sensitive = true
}

data "external" "this_works" {
  program = ["jq", "."]

  query = {
    # arbitrary map from strings to strings, passed
    # to the external program as the data query.
    secret = var.secret
  }
}

resource "null_resource" "this_errors" {
  provisioner "local-exec" {
    interpreter = ["/bin/bash", "-c"]
    command     = "jq ."
    environment = {
      SECRET = var.secret
    }
  }
}

resource "null_resource" "this_works" {
  provisioner "local-exec" {
    interpreter = ["/bin/bash", "-c"]
    command     = "jq ."
    environment = {
      SECRET = format("%s", var.secret)
    }
  }
}

```

### Expected Behavior

A `terraform plan` should not error but the secret should remain redacted.

### Actual Behavior

I get an error message:
```
Error: .environment["SECRET"]: value has marks, so it cannot be serialized
```

### Steps to Reproduce

- Create terraform file as above.
- `terraform init && terraform plan`

### Additional Context

This is reproducible with terraform as vanilla as you can get.

### References
- #27095

**Comment by oliverisaac** [comment]
As a workaround for this issue, you can pass the variable through `format("%s", var.secret)` and it will not error (as show in the example.

**Comment by apparentlymart** [comment]
Thanks for reporting this, @oliverisaac.

Although I agree that this error message seems to be incorrect (this particular error message is reporting an internal problem rather than a user error), I wanted to also note that the workaround you used of passing the sensitive value through `format` to remove the sensitivity seems to be indicative of a separate bug in the `format` function, where it's failing to mark its result as sensitive when one of the inputs is sensitive.

**Comment by oliverisaac** [comment]
> seems to be indicative of a separate bug in the format function, where it's failing to mark its result as sensitive when one of the inputs is sensitive.

I don't think  this is the case. If you use this test case:

```hcl
variable "secret" {
  default   = "password1235"
  sensitive = true
}

resource "null_resource" "this_works" {
  provisioner "local-exec" {
    interpreter = ["/bin/bash", "-c"]
    command     = "jq ."
    environment = {
      SECRET = format("%s", var.secret)
    }
  }

  triggers = {
    secret = format("%s", var.secret)
  }
}
```

And then run this multiple times, you'll see that it marks the value as sensitive in both the plan and apply:

```bash
TF_VAR_secret=$( date +%s ) terraform apply <<< yes
```

<img width="1060" alt="Screen Shot 2021-02-14 at 12 44 05 PM" src="https://user-images.githubusercontent.com/28768356/107885898-69a40600-6ec2-11eb-81e2-a052c50d50af.png">

**Comment by alisdair** [comment]
Thanks for the example configuration! I was able to confirm this issue on 0.14.7. It seems to be fixed on the latest build from the main branch, but I'm not sure what could have changed that would have caused this.

**Comment by alisdair** [comment]
The error is (perhaps predictably) surfacing when we attempt to validate the provisioner config, here: https://github.com/hashicorp/terraform/blob/74134a8f30a54cd3479dfeeb6d2822ac56609619/terraform/eval_validate.go#L96-L100

There has been some significant refactoring in the main branch, so this code is now here: https://github.com/hashicorp/terraform/blob/48bfde2e28b988b2f48b6cc0129719929e0694b1/terraform/node_resource_validate.go#L93-L97

The reason this bug is not reproducible with the main branch is because as of #27077 the built-in provisioners run in-process, rather than through the plugin interface, so we don't need to marshal the configuration value. I still think this is a valid bug that we can fix, but writing tests for this on the main branch will be a bit harder.

**Comment by alisdair** [comment]
There's a fix pending review in #27819. With a build from that branch, your original configuration now succeeds:

```shellsession
$ terraform apply

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # null_resource.this_errors will be created
  + resource "null_resource" "this_errors" {
      + id = (known after apply)
    }

  # null_resource.this_works will be created
  + resource "null_resource" "this_works" {
      + id = (known after apply)
    }

Plan: 2 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

null_resource.this_errors: Creating...
null_resource.this_works: Creating...
null_resource.this_errors: Provisioning with 'local-exec'...
null_resource.this_works: Provisioning with 'local-exec'...
null_resource.this_errors (local-exec): (output suppressed due to sensitive value in config)
null_resource.this_works (local-exec): (output suppressed due to sensitive value in config)
null_resource.this_works: Creation complete after 0s [id=6126156555107701453]
null_resource.this_errors: Creation complete after 0s [id=627371408651545635]

Apply complete! Resources: 2 added, 0 changed, 0 destroyed.
```

Thanks again for such a thorough bug report!

One final mystery is what's going on with the `format` function described in your workaround. I looked at that briefly but I really have no idea what's happening there.

**Comment by pselle** [comment]
#27819 has been merged, closing this (not sure why GitHub didn't :) )
